### ISSUE TITLE
Mounting Bugs

### ISSUE THREAD

**Issue by jbenet** [jbenet/go-ipfs#574]
cc @whyrusleeping @briantigerchow @mappum -- this concerns you all.

There's a couple major mounting bugs, which may be related:

1. [x] Mounting now hangs. On OSX, mounting with the daemon on hangs. Not sure why-- it seems to be waiting on some sort of network IO. It _may_ be related to ipns, as `/ipfs` finishes mounting well before the `ipfs mount` command returns, but only sometimes.

2. [ ] `ipfs mount` only works with the daemon on. (UX flaw) Previously, we had resolved to only enable `ipfs mount` when the daemon is on, but if the network is chatty, or the user finds herself in a poor network condition, running all online services may not be desired.

(1) may be fixed simply, if we find what net IO it's doing.
(2) allowing `ipfs mount` to run without the daemon introduces complexity on the core + daemon. From a UX point of view, this should be possible:

```sh
> ipfs init
> ipfs add -r foo
added QmTz3oc4gdpRMKP2sdGUPZTAGRngqjsi99BPoztyP53JMM test/bar
added QmTz3oc4gdpRMKP2sdGUPZTAGRngqjsi99BPoztyP53JMM test/baz/b
added QmYNmQKp6SuaVrpgWRsPTgCQCnpxUYGq76YEKBXuj2N4H6 test/baz/f
added QmX1ebVUtfY11ZCpVmqyE5mDoN62SpLd8eLPpg5GGV1ABt test/baz
added QmYNmQKp6SuaVrpgWRsPTgCQCnpxUYGq76YEKBXuj2N4H6 test/foo
added QmNtpA5TBNqHrKf3cLQ1AiUKXiE4JmUodbG5gXrajg8wdv test
> ipfs mount &  # whether to background or whatever is left to implementation.
IPFS mounted at /ipfs
IPNS mounted at /ipns
> ipfs cd /ipfs/QmNtpA5TBNqHrKf3cLQ1AiUKXiE4JmUodbG5gXrajg8wdv
> ls
bar
baz/
foo
```

Presently, we do the simple thing for us (force the daemon to be run first), which introduces undue complexity _on the user_. The user must now know about ipfs daemons and online servies. And the user may incurr the cost of doing network IO. 

We can go about this in many ways, but 3 options stand out for me. Please suggest others if you see them! These are discussed at length below. 

1. Keep `ipfs mount` in the daemon's domain, but make the dameon able to run offline (discussed below).
2. Make "`ipfs mount` with no running daemon" mount the FS until exit, blocking any subsequent calls to `ipfs daemon` or `ipfs mount`. (discussed below)
3. Make `ipfs mount` _always_ be a long running process.

 ## 1.  online/offline daemon

For `ipfs mount` to require running `ipfs daemon` and not _force_ network IO on the user, `ipfs daemon` must learn two modes: `{Online, Offline}`. Whether transitioning from `Offline` to `Online` requires rebooting the process, or it can achieve it without rebooting[2], `ipfs mount` could check for a daemon, if none is running, start it with `--offline`, and then proceed as usual. If `ipfs unmount`[3] is called, the offline daemon could realize it is no longer needed, and exit. we must be careful not to leave a daemon running without the user's desire. i.e. this should NOT leave resources hanging:

```sh
> ipfs init
> ipfs mount  # may kick off a daemon
> ipfs unmount
# at this point, no daemon should be left running indefinitely.
# it could be in the process of exiting.
```

Having `{Online/Offline}` modes on the daemon moves some of the complexity of other commands (i.e. when should a daemon be run offline vs online), but may actually simplify other uses[4]. This might also make other features available: manage the node, or browse local files from the webui _without being online and connected to the broader ipfs network_.

[1] starting with `ipfs daemon` vs `ipfs daemon --offline` or something equivalent (`--online={true,false}` defaulting to true?)

[2] can toggle this from daemon API clients. e.g. from `cli`, by running `ipfs daemon [ connect | disconnect ]` or rsomething similar, and from `webui` by clicking a button.

[3] i'd want both the new-user-friendly `ipfs unmount` and the traditional-user-friendly `ipfs umount` to work. We dont have to list both, one can be a simple alias of the other.

[4] we _could_ make `ipfs` always use a daemon, offline or online, and launch it if need be, garbage collecting when it's done. This would _seriously_ simplify implementation for us. We could distinguish between two binaries: `{ipfs, ipfsd}` as is traditional with many daemonized systems. This would have the advantage of being what most unix users might expect. This has the disadvantage of (maybe) having to launch 2 processes and do RPC for _every single_ `ipfs` command.

## 2. "`ipfs mount` with no daemon" mounts the FS until exit

This would be the trivial fix today. `ipfs mount` checks if there is a daemon running, if not, it mounts the FS until exit. This blocks the user from running `ipfs daemon`, as it should lock the repo (it uses the datastore):

```sh
> ipfs init
> ipfs mount &
> ipfs daemon
Error: repository locked by process <pid>. You may need to stop it to run this command.
```

Advantages: 
- super trivial fix

Disadvantages:
- complicates the user's life: is `ipfs mount` running? is `ipfs daemon` running? why does `ipfs mount` sometimes run fuse, and sometimes mount through the daemon?
- having not one, but TWO types of repo-holding long-running process increases the complexity of all repo-related use.

### 3. `ipfs mount` is _always_ long-running.

This is like (2), but actually takes over mounting entirely from the daemon. The way it would work is that `ipfs mount` is always a long-running process, and does the fuse mounting itself. (and thus must be backgrounded). BUT! it's smart, whenever an object is requested, it does the same thing that the commands interface would do: check if there is a daemon, if yes, use it. if not, use repo directly, as `ipfs-cli` would normally. This should work:

```sh
> ipfs mount &
> cat /ipfs/<hash>   # backgrounded mount opens repo, cats, closes repo
> ipfs daemon &
> cat /ipfs/<hash>   # backgrounded mount uses daemon
```

This has one _major_ advantage. It would be possible to use `ipfs mount` to do temporary mounts of only specific paths in the merkledag: 

```sh
> ls
> ipfs mount QmNtpA5TBNqHrKf3cLQ1AiUKXiE4JmUodbG5gXrajg8wdv test
> ls
test/
> ls test/
bar
baz/
foo
```
(actually, this is a good idea, we should probably do it no matter which of these 3 approaches we take. It makes it really easy to build small unix tools that use only a small portion of the ipfs merkledag in a unix filesystem friendly way).

Advantages:
- This is _the right way_ to do (2)
- unambigious (2), easier on the user: `ipfs mount` _always_ behaves the same way
- provides a _really good_ abstraction of pieces (mount and daemon are two separate things)

Disadvantages:
- may be very slow.
- user must now manage two kinds of long running processes: daemon's and mounts.

## Which way fwd?

Not sure what the _right_ way forward is. (1) is certainly harder from where we are, but seems like TRTTD. (2) is today's easy hack. it may introduce a _ton_ of complexity later, if we choose to keep this as a mode. We could also opt to do (1) down the road, but implement (2) now as a stopgap, with the full intention of changing to (1) and abandoning (2).

Thoughts?

**Comment by whyrusleeping** [comment]
Mounting appears to hang for me too, on linux. Which is very strange since the sharness tests pass

**Comment by jbenet** [comment]
yeah i think its somehow waiting for IO. what does ipns do on init? does it
do any lookups?

**Comment by whyrusleeping** [comment]
ooooh, it does do a name resolution. I can look into it if youd like:

https://github.com/jbenet/go-ipfs/blob/master/fuse/ipns/ipns_unix.go#L89

**Comment by whyrusleeping** [comment]
Ah yeah, we need to pass a context into `Resolve` its currently running a `GetValue` query with a `context.TODO()` and hanging... Not sure why its hanging, dht queries should eventually halt on their own.

**Comment by jbenet** [comment]
> Ah yeah, we need to pass a context into `Resolve` its currently running a GetValue query with a `context.TODO()` and hanging

Any chance that `Resolve` can be async? or does it have to be _finished_ before returning a successful/unsuccessful mount? 

> dht queries should eventually halt on their own.

only after querying lots of the network...
And, it's not halting, just taking around 30s-1m

**Comment by whyrusleeping** [comment]
So, what its doing, is for each key, it needs to resolve the ipfs entry it currently points to. It uses this to populate the initial ipns filesystem. These values *should* be stored locally once created, but i think that what we are running into, is that there arent any values for the key to resolve to. One easy solution would be to make your key point to an 'empty dir' upon `ipfs init`

**Comment by jbenet** [comment]
> for each key, it needs to resolve the ipfs entry it currently points to. It uses this to populate the initial ipns filesystem.

hmm and does it _need_ to happen at init? why not when `ls` ing `/ipns ?

> One easy solution would be to make your key point to an 'empty dir' upon `ipfs init`

SGTM! 

though, exactly _what_ happens in this init should be part of ipns packages, and the core can just make sure to call it.

**Comment by whyrusleeping** [comment]
Oh yeah, we could have a function `ipns.Initialize(keys)` that would make each of the keys you give it point to an empty dir (stored locally)

**Comment by jbenet** [comment]
@briantigerchow i'd love your thoughts on the second problem listed here: 

> `ipfs mount` only works with the daemon on ...

**Comment by whyrusleeping** [comment]
added those tags to the wrong issue.

**Comment by whyrusleeping** [comment]
@jbenet Is the second issue discussed on the alpha TODO list?

**Comment by jbenet** [comment]
@whyrusleeping yep

**Comment by whyrusleeping** [comment]
My thoughts are that we should have the current mounting semantics, whether offline or online, this calls for an 'offline' mode for the daemon.

```
ipfs daemon -offline &
ipfs mount
```

I think in general we do want an offline daemon, so work towards that end will be helpful in any case.

**Comment by whyrusleeping** [comment]
Also, as you talk about having `ipfs daemon { connect | disconnect }` to modify the live running state of the daemon would be really nice.

**Comment by pataquets** [comment]
Is there currently any way to make a node started with ```--offline``` switch to go online? None of the ```ipfs config show``` settings printed looks suitable. Googling also failed, so if it is possible to do so, the procedure is a good candidate for docs/faqs (advice on where to do it myself also welcome).

**Comment by whyrusleeping** [comment]
@pataquets Thats what i'm proposing in this issue: https://github.com/ipfs/go-ipfs/issues/3372
